HW2/P2/parallel_vector.pyx

```diff
@@ -0,0 +1,154 @@
+##################################################
+
+
+from cython.parallel import parallel, prange
+from openmp cimport omp_lock_t, \
+    omp_init_lock, omp_destroy_lock, \
+    omp_set_lock, omp_unset_lock, omp_get_thread_num
+from libc.stdlib cimport malloc, free
+
+import numpy as np
+cimport numpy as np
+
+
+# lock helper functions
+cdef void acquire(omp_lock_t * l) nogil:
+    omp_set_lock(l)
+
+cdef void release(omp_lock_t * l) nogil:
+    omp_unset_lock(l)
+
+# helper function to fetch and initialize several locks
+cdef omp_lock_t * get_N_locks(int N) nogil:
+    cdef:
+        omp_lock_t * locks = <omp_lock_t * > malloc(N * sizeof(omp_lock_t))
+        int idx
+
+    if not locks:
+        with gil:
+            raise MemoryError()
+    for idx in range(N):
+        omp_init_lock(& (locks[idx]))
+
+    return locks
+
+cdef void free_N_locks(int N, omp_lock_t * locks) nogil:
+    cdef int idx
+
+    for idx in range(N):
+        omp_destroy_lock(& (locks[idx]))
+
+    free( < void * > locks)
+
+
+##################################################
+# Your code below
+##################################################
+
+cpdef move_data_serial(np.int32_t[:] counts,
+                       np.int32_t[:] src,
+                       np.int32_t[:] dest,
+                       int repeat):
+    cdef:
+        int idx, r
+
+    assert src.size == dest.size, "Sizes of src and dest arrays must match"
+    with nogil:
+        for r in range(repeat):
+            for idx in range(src.shape[0]):
+                if counts[src[idx]] > 0:
+                    counts[dest[idx]] += 1
+                    counts[src[idx]] -= 1
+
+
+cpdef move_data_fine_grained(np.int32_t[:] counts,
+                             np.int32_t[:] src,
+                             np.int32_t[:] dest,
+                             int repeat):
+    cdef:
+        int idx, r
+        omp_lock_t * locks = get_N_locks(counts.shape[0])
+
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for each element of counts to parallelize
+    # data movement.  Be sure to avoid deadlock, and double-locking.
+    ##########
+    with nogil:
+        for r in prange(repeat, num_threads = 4):
+            for idx in range(src.shape[0]):
+
+                if counts[src[idx]] > 0:
+                    
+                    #We lock the elements corresponding to the destination and source indexes. We always lock the smallest index before to avoid deadlocks.
+                    if dest[idx] < src[idx]:
+                        acquire(&(locks[dest[idx]]))
+                        acquire(&(locks[src[idx]]))
+                    elif dest[idx] > src[idx]:
+                        acquire(&(locks[src[idx]]))
+                        acquire(&(locks[dest[idx]]))
+                    else:
+                        #In case the indexes are the same we just lock it once to avoid double locking.
+                        acquire(&(locks[dest[idx]]))
+
```

Comment:
Nice, I implemented my locking scheme just as you did, great minds think alike ;)